### ISSUE TITLE
GDPR non-compliance regarding consent and/or personal data processing

### ISSUE THREAD

**Issue by KenVanHoeylandt** [Homebrew/brew#11662]
### What did you expect to happen?

One of the scenarios below:

1. Make analytics tracking opt-in.
2. Remove the usage of a unique user id.

(there might be more options)

### Step-by-step reproduction instructions (by running `brew` commands)

1. Observe how [analytics.sh](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/analytics.sh) uses `/usr/bin/uuidgen` to make a unique user id.

2. Note that GDPR [considers](https://gdpr.eu/article-4-definitions/) such unique id as "personal data": 

> ‘personal data’ means any information relating to an identified or identifiable natural person (‘data subject’); an identifiable natural person is one who can be identified, directly or indirectly, in particular by reference to an identifier such as a name, **an identification number**, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person;

(emphasis added)

3. [Consent](https://gdpr.eu/article-4-definitions/) through opt-out is not considered as a valid consent to process personal data:

> ‘consent’ of the data subject means any freely given, specific, informed and **unambiguous indication** of the data subject’s wishes by which he or she, by a statement or by a clear affirmative action, signifies agreement to the processing of personal data relating to him or her;

(emphasis added)

Additional source giving opt-in for a marketing mailing list as an example https://gdpr.eu/what-is-gdpr/

4. Considering that personal data (uuid) is being processed by a third party (Google), I believe that the consent must be given and also be explicit (opt-in). This can be mitigated by either not processing personal data or having an opt-in mechanism.

**Comment by carlocab** [comment]
> > ‘personal data’ means any information relating to an identified or identifiable natural person (‘data subject’); an identifiable natural person is one who can be identified, directly or indirectly, in particular by reference to an identifier such as a name, **an identification number**, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person;
> 
> (emphasis added)

Can you be identified from your analytics UUID?

From https://gdpr.eu/what-is-gdpr/:

>Personal data — Personal data is any information that relates to an individual who can be directly or indirectly identified. Names and email addresses are obviously personal data. Location information, ethnicity, gender, biometric data, religious beliefs, web cookies, and political opinions can also be personal data. Pseudonymous data can also fall under the definition if it’s relatively easy to ID someone from it.

This tells me that data doesn't become personal data just because it's tagged with a UUID. It could be, but we probably need to establish how easy it is to identify a user from their analytics UUID.

**Comment by KenVanHoeylandt** [comment]
@carlocab The "unique user identification" directly identifies to you as user, as is implied by the term itself. Whether it can be linked to other personal data (such as your name or location) is not relevant in the context of determining whether a uuid is considered personal data.

Note that the "What is GDPR" page is only there to help to explain the topics on a higher level and with some examples. It doesn't cover all the flavours of technology. The actual law articles that I linked are more relevant.

**Comment by KenVanHoeylandt** [comment]
To clarify that uuid is considered personal data, I will break it down from the GDPR law article text:

> ‘personal data’ means any information relating to an identified or identifiable natural person (‘data subject’); an identifiable natural person is one who can be identified, directly or indirectly, in particular by reference to an identifier such as a name, an identification number, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person;

Looking at the first part:

> ‘personal data’ means any information relating to an identified or identifiable natural person (‘data subject’);

We can confirm that uuid does indeed relate to an identified or identifiable natural person. It relates in this case to a user of a computer.

Edge case: In theory, a computer user can be shared among multiple natural persons, but for all intents and purposes, a computer user generally relates to (and is intended to relate to) a single natural person with an account on a computer.
Even if some cases will be of shared computer usage, it is safe to assume that many (if not most) computer users will relate to a specific natural person.

> an identifiable natural person is one who can be identified, directly or indirectly, in particular by reference to an identifier such as a name, an identification number, location data, an online identifier or to one or more factors specific to the physical, physiological, genetic, mental, economic, cultural or social identity of that natural person;

We can confirm that the user is being directly identified by means of "an identification number". As the name even says, this number even is intended to uniquely identify a user.
I interpret that the identification is not indirect, as we don't need an aggregation of other data points to create an indirect reference to the user: the uuid has the intention to create a 1-to-1 relationship to the computer user.

**Comment by carlocab** [comment]
It seems to me that there is ambiguity in what it means for an individual to be "directly or indirectly identified". I think your interpretation of what "identified" means implies that all pseudonymous data must be construed as "personal data", which isn't consistent with the "What is GDPR" page. Is that page wrong?

**Comment by KenVanHoeylandt** [comment]
Ah, I missed that. Thanks!

From what I understand, the remaining question is then: can I be identified by the remaining parameters?
I did some searching:

IP anonymization as applied by Google Analytics is not as anonymous as it seems: https://support.google.com/analytics/answer/2763052?hl=en 
With IPV4, it only removes the last octet. This means that 111.112.113.114 becomes 111.112.113.0 for every user in the IP range 111.112.113.0-111.112.113.255
(note that it is also documented that this data is used to identify you geographically!)

To de-anonymize the (v4) IP address, we only need to know a number in the range of [0, 255]. So we only need enough data to identify a person from a group of (at most!) 256 other persons.

I'm convinced that if we take all active(!) IP addresses from this range of maximum 256 in total, I'm sure there's only 1 person with my user agent (e.g. `Homebrew/2.5.0 (Macintosh; Intel Mac OS X 10.15.6) curl/7.64.1`) and my specific list of installed packages (implied from the events that are being sent).
The list of packages is likely not even necessary to identify a user in many cases.

In conclusion: I am still convinced that a Brew user can be identified.

**Comment by carlocab** [comment]
There are two questions here.

1. Is the information that Homebrew collects personal data, as in the language of the GDPR?
2. Assuming it is personal data, does the GDPR require that Homebrew make the analytics opt-in rather than opt-out?

Your contention is that the information that Homebrew collects is sufficient to identify an individual. I'm not sure I follow. Suppose I gave you a Homebrew analytics UUID, along with the associated user agent, and a full, non-anonymous IP address. [1] How would you use this information to identify (i.e. name) the associated user?

I would think that this is not enough to identify an individual. At the very least, doing so would be costly and time-consuming [2], even if hypothetically you had access to time-series data associated with a given analytics UUID. It seems to me that the answer then to the first question should be no.

On the second question:

The primary requirement of the GDPR appears to be

> Personal data shall be:
> - processed lawfully, fairly and in a transparent manner in relation to the data subject (‘lawfulness, fairness and transparency’);

In particular, the requirement about consent pertains to the "lawfulness" requirement, as we see in [Article 6](https://gdpr.eu/article-6-how-to-process-personal-data-legally/), which I quote selectively below (emphasis mine):

> Processing shall be lawful only if and to the extent that **at least one of** the following applies:
>
> a. the data subject has given consent to the processing of his or her personal data for one or more specific purposes;
> ...
> f. processing is necessary for the purposes of the legitimate interests pursued by the controller or by a third party, except where such interests are overridden by the interests or fundamental rights and freedoms of the data subject which require protection of personal data, in particular where the data subject is a child.

So, even assuming the analytics data were personal data (it's not), the GDPR does not necessarily require that they must be collected only with opt-in consent. I would argue that processing of data (personal or otherwise) is lawful under Article 6 1.f. (See the link in [1] for reasons why.)

[1] This is actually much more information than what Homebrew has access to. See https://docs.brew.sh/Analytics.
[2] [Recital 26](https://gdpr.eu/Recital-26-Not-applicable-to-anonymous-data) looks applicable here.

**Comment by iMichka** [comment]
I think there are other topics discussing GPDR in homebrew in our different repositories.
The conclusion from the maintainers was that we were not doing anything wrong, and we would change nothing to the way we handle things.

As there is nothing actionable here, I will close this issue. We do not consider this an issue or a bug.

Of course if you want to continue discussing this you could open a thread at https://github.com/Homebrew/discussions.
The only actionable thing I could see would be to pay an actual GPDR expert to make a statement that everything is fine, but for that we would need to check with the homebrew PLC if that expense makes sense.

**Comment by MikeMcQuaid** [comment]
Homebrew's license pretty clearly waives all warranties and liabilities: https://github.com/Homebrew/brew/blob/master/LICENSE.txt

You agree to this license when you use Homebrew.

It's very much unclear to me how GDPR somehow overrides this license.

**Comment by KenVanHoeylandt** [comment]
@MikeMcQuaid Licenses (and contracts) don't override legal protections.
e.g. If someone writes in a license that by using the software, you'd have to give your first-born child to the developer, that won't make it legal.

In the case of GDPR, consent is not considered as "freely given" when the user is forced to hand over their data to use a product/service - with the exception of data that is necessary to perform that contract.

An example of this are the "cookie dialogs" that you see on many websites. Some websites refuse you access if you deny non-functional cookies. The latter is against GDPR regulations.

In the case of Homebrew: Sending analytics data to a third party is not a necessity for Homebrew's core functionality, so that waiver of the license wouldn't apply to this scenario.

[Article 7 point 4](https://gdpr.eu/article-7-how-to-get-consent-to-collect-personal-data/)):
> When assessing whether consent is freely given, utmost account shall be taken of whether, inter alia, the performance of a contract, including the provision of a service, is conditional on consent to the processing of personal data that is not necessary for the performance of that contract.

**Comment by MikeMcQuaid** [comment]
> Licenses (and contracts) don't override legal protections.

Neither of us are lawyers so I don't think either of us can state either way. Similarly, GDPR is an EU law and I'm (sadly) not in an EU country. Homebrew is part of the SFC which is also not in an EU country.

> In the case of GDPR, consent is not considered as "freely given" when the user is forced to hand over their data to use a product/service - with the exception of data that is necessary to perform that contract.

Homebrew is not a product or service. It is software that you can use on your machine if you agree with the license. If you believe that license does not comply with GDPR: do not use Homebrew.

> Sending analytics data to a third party is not a necessity for Homebrew's core functionality

It is necessary to run the Homebrew project.

---

What would be a more helpful use of your time would be finding ways of further anonymising our analytics or finding a GDPR-compliant analytics host for us.

**Comment by carlocab** [comment]
Even if you don't think the license has any bearing wrt GDPR:

1. The data that Homebrew collects is not personal data. See my previous comment for an explanation. Note in particular [recital 26](https://gdpr.eu/Recital-26-Not-applicable-to-anonymous-data), which states we must account for temporal and technological constraints when determining whether collected data can be used to identify a person.
2. Even if the data did constitute personal data (it does not), the GDPR does not necessarily mandate that users must explicitly consent to the collection of this data. Homebrew would be unmaintainable if we had no analytics data, or if we only had the analytics data from users who opt-in. Thus, the collection of this data is lawful under GDPR 6.1.f.

**Comment by KenVanHoeylandt** [comment]
@carlocab You have good points and I wasn't arguing against yours. I appreciate your input and efforts! MikeMcQuaid was implying that the license overrides the GDPR, which is false.

**Comment by carlocab** [comment]
Sure. I was just hoping you had an answer to my question:

If I gave you an analytics UUID, along with the associated user agent and full (!) IP address, would you be able to tell me who this user was? [1] Or, at least, devise a procedure that would allow you to? You seemed to claim earlier that you could, but I don't quite see how. [2]

Your answer to the above would help us determine whether (1) the data Homebrew collects constitutes personal data, and (2) there are other things Homebrew can do to make sure the collected data is actually anonymous and therefore would not be personal data.

[1] This is actually purely hypothetical. I (or any other Homebrew maintainer) can't actually see individual UUIDs and associated user agents or (anonymised) IP addresses. [All we see are numbers of install events for each formula broken down by OS](https://docs.brew.sh/Analytics#what).

[2] I can see that the NSA probably could use this data to identify a person, but I think this is where the mention of technological and temporal constraints in recital 26 kicks in.

**Comment by MikeMcQuaid** [comment]
> You have good points and I wasn't arguing against yours.

@KenVanHoeylandt Apparently you have no desire to reply to mine. 

> MikeMcQuaid was implying that the license overrides the GDPR, which is false.

I'm taking from the lack of disagreement where I said you aren't a lawyer that you aren't. I'm not either. Neither is @carlocab. As a result, despite @carlocab's best efforts to provide helpful information and my best efforts to point out how you can actually make changes to help make things better: you have no desire to do so. Locking.
